ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,536 @@
+     * */
+    public fun beforeHandle(block: suspend CallExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Defines what to do with the call in case request was not handled for some reason.
+     * Usually, it is handy to use fallback { call -> ...} to set the response with some message or status code,
+     * or to throw an exception.
+     * */
+    public fun unhandled(block: suspend CallExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface OnReceive {
+    /**
+     * Define how current [KtorPlugin] should transform data received from a client.
+     * */
+    public operator fun invoke(block: suspend ReceiveExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Defines actions to perform before any transformations were made to the received content. It is useful for caching.
+     * */
+    public fun beforeTransform(block: suspend ReceiveExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface OnRespond {
+    /**
+     * Do transformations of the data. Example: you can write a custom serializer using this method.
+     * (Note: it is handy to also use [Execution.proceedWith] for this scenario)
+     * */
+    public operator fun invoke(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Allows to use the direct result of call processing (see [OnRequest.invoke]) and prepare data before sending a response will be executed.
+     * */
+    public fun beforeTransform(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Allows to calculate some statistics on the data that was already sent to a client, or to handle errors.
+     * (See [Metrics], [CachingHeaders], [StatusPages] features as examples).
+     * */
+    public fun afterTransform(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface PluginContext {
+    public val onRequest: OnRequest
+    public val onCallReceive: OnReceive
+    public val onCallRespond: OnRespond
+}
+
+/**
+ * Gets plugin instance for this pipeline, or fails with [MissingApplicationFeatureException] if the feature is not installed
+ * @throws MissingApplicationFeatureException
+ * @param plugin plugin to lookup
+ * @return an instance of plugin
+ */
+public fun <A : Pipeline<*, ApplicationCall>> A.plugin(
+    plugin: KtorInstallablePlugin<Configuration>
+): KtorPlugin<Configuration> {
+    return attributes[featureRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationFeatureException(plugin.key)
+}
+
+internal fun <A : Pipeline<*, ApplicationCall>> A.findInterceptionsHolder(
+    plugin: KtorInstallablePlugin<*>
+): InterceptionsHolder {
+    return attributes[featureRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationFeatureException(plugin.key)
+}
+
+/**
+ * [Execution] holds everything that is bound to the current plugin execution. Instance of [Execution] can be
```

Comment:
You can't explain `Execution` through `execution`, it's not clear what `plugin execution` is.